blocks/graph.py

```diff
@@ -354,3 +386,45 @@ def apply_noise(computation_graph, variables, level, seed=None):
         replace[variable] = (variable +
                              rng.normal(variable.shape, std=level))
     return computation_graph.replace(replace)
+
+
+def apply_dropout(computation_graph, variables, drop_prob=0.5, rng=None,
+                  seed=None):
+    """Returns a graph to variables in a computational graph.
+
+    Parameters
+    ----------
+    computation_graph : instance of :class:`ComputationGraph`
+        The computation graph.
+    variables : list of :class:`~tensor.TensorVariable`
+        Variables to be dropped out.
+    drop_prob : float
+        Probability of dropping out. If you want to apply the dropout
+        with different probabilities for different layers, call it
+        several times.
+    rng : :class:`~theano.sandbox.rng_mrg.MRG_RandomStreams`
+        Random number generator.
+    seed : int
+        Random seed to be used if `rng` was not specified.
+
+    Notes
+    -----
+    For more information, see [DROPOUT]_.
+
+    .. [DROPOUT] Hinton et al. *Improving neural networks by preventing
+       co-adaptation of feature detectors*, arXiv:1207.0580.
+
+    """
+    if not rng and not seed:
+        seed = config.default_seed
+    if not rng:
+        rng = MRG_RandomStreams(seed)
+
+    replacements = [(var, var * rng.binomial(var.shape, p=1 - drop_prob,
+                                             dtype=floatX) / (1 - drop_prob))
+                    for var in variables]
+    for variable, replacement in replacements:
+        replacement.tag.roles = variable.tag.roles + [DROPOUT]
```

Comment:
Yeah, that's what we settled on. So `replacement.add_role(DROPOUT)` should suffice here. The rest of the roles can be found by following the `replacement_of` link.